### PR TITLE
add netrc support for webengine. resolves #2785.

### LINKED ISSUE #2785: netrc support for QtWebEngine
Abandoned PR from #2594:

```patch
From 09c6c4d1d6b4a020669df23e9be9043dce799d50 Mon Sep 17 00:00:00 2001
From: Kevin Velghe <kevin@paretje.be>
Date: Fri, 28 Apr 2017 08:43:44 +0200
Subject: [PATCH] Implement netrc for webengine

---
 qutebrowser/browser/shared.py                      | 29 ++++++++++++++++++++++
 qutebrowser/browser/webengine/webenginetab.py      | 11 ++++----
 .../browser/webkit/network/networkmanager.py       | 23 +----------------
 3 files changed, 36 insertions(+), 27 deletions(-)

diff --git a/qutebrowser/browser/shared.py b/qutebrowser/browser/shared.py
index 885e2809d1..f515f77abe 100644
--- a/qutebrowser/browser/shared.py
+++ b/qutebrowser/browser/shared.py
@@ -23,6 +23,9 @@
 
 import jinja2
 
+import os
+import netrc
+
 from qutebrowser.config import config
 from qutebrowser.utils import usertypes, message, log, objreg
 
@@ -243,3 +246,29 @@ def get_user_stylesheet():
         css += '\nhtml > ::-webkit-scrollbar { width: 0px; height: 0px; }'
 
     return css
+
+
+def netrc_authentication(url, authenticator):
+    if 'HOME' in os.environ:
+        # We'll get an OSError by netrc if 'HOME' isn't available in
+        # os.environ. We don't want to log that, so we prevent it
+        # altogether.
+        user, password = None, None
+        try:
+            net = netrc.netrc(config.get('network', 'netrc-file'))
+            authenticators = net.authenticators(url.host())
+            if authenticators is not None:
+                (user, _account, password) = authenticators
+        except FileNotFoundError:
+            log.misc.debug("No .netrc file found")
+        except OSError:
+            log.misc.exception("Unable to read the netrc file")
+        except netrc.NetrcParseError:
+            log.misc.exception("Error when parsing the netrc file")
+
+        if user is not None:
+            authenticator.setUser(user)
+            authenticator.setPassword(password)
+            return True
+
+    return False
diff --git a/qutebrowser/browser/webengine/webenginetab.py b/qutebrowser/browser/webengine/webenginetab.py
index cedd59b1ed..1e8e31c5e2 100644
--- a/qutebrowser/browser/webengine/webenginetab.py
+++ b/qutebrowser/browser/webengine/webenginetab.py
@@ -631,11 +631,12 @@ def _on_history_trigger(self):
 
     @pyqtSlot(QUrl, 'QAuthenticator*')
     def _on_authentication_required(self, url, authenticator):
-        # FIXME:qtwebengine support .netrc
-        answer = shared.authentication_required(
-            url, authenticator, abort_on=[self.shutting_down,
-                                          self.load_started])
-        if answer is None:
+        netrc = shared.netrc_authentication(url, authenticator)
+        if not netrc:
+            answer = shared.authentication_required(
+                url, authenticator, abort_on=[self.shutting_down,
+                                              self.load_started])
+        if not netrc and answer is None:
             try:
                 # pylint: disable=no-member, useless-suppression
                 sip.assign(authenticator, QAuthenticator())
diff --git a/qutebrowser/browser/webkit/network/networkmanager.py b/qutebrowser/browser/webkit/network/networkmanager.py
index 779d778bce..c32d6d47e2 100644
--- a/qutebrowser/browser/webkit/network/networkmanager.py
+++ b/qutebrowser/browser/webkit/network/networkmanager.py
@@ -277,28 +277,7 @@ def clear_rejected_ssl_errors(self, url):
     @pyqtSlot('QNetworkReply*', 'QAuthenticator*')
     def on_authentication_required(self, reply, authenticator):
         """Called when a website needs authentication."""
-        user, password = None, None
-        if not hasattr(reply, "netrc_used") and 'HOME' in os.environ:
-            # We'll get an OSError by netrc if 'HOME' isn't available in
-            # os.environ. We don't want to log that, so we prevent it
-            # altogether.
-            reply.netrc_used = True
-            try:
-                net = netrc.netrc(config.get('network', 'netrc-file'))
-                authenticators = net.authenticators(reply.url().host())
-                if authenticators is not None:
-                    (user, _account, password) = authenticators
-            except FileNotFoundError:
-                log.misc.debug("No .netrc file found")
-            except OSError:
-                log.misc.exception("Unable to read the netrc file")
-            except netrc.NetrcParseError:
-                log.misc.exception("Error when parsing the netrc file")
-
-        if user is not None:
-            authenticator.setUser(user)
-            authenticator.setPassword(password)
-        else:
+        if not shared.netrc_authentication(reply.url(), authenticator):
             abort_on = self._get_abort_signals(reply)
             shared.authentication_required(reply.url(), authenticator,
                                            abort_on=abort_on)
```

Unfortunately this gets us in an endless loop with a wrong password...

### PR DESCRIPTION
**Pull request by mjauvin**
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3505)

---

Resolves #2785. *(edited by @The-Compiler as issue references need to be in the body to work on GitHub)*

### CODE REVIEW

**Comment by The-Compiler** [comment]
You never reset `netrc_used` anywhere, which means that it'll only work once per tab - right?

**Comment by mjauvin** [comment]
@The-Compiler netrc_used is reset in _on_load_started() in AbstractTab class.

I've tested this and it works. Do you have a case where it would fail?

File: qutebrowser/browser/webengine/webenginetab.py [hunk 1]
```diff
@@ -741,7 +741,7 @@ def _on_proxy_authentication_required(self, url, authenticator,
     @pyqtSlot(QUrl, 'QAuthenticator*')
     def _on_authentication_required(self, url, authenticator):
         netrc = None
-        if not self.data.netrc_used and 'HOME' in os.environ:
+        if not self.data.netrc_used:
```
Comment (mjauvin) [inline qutebrowser/browser/webengine/webenginetab.py]: ok, this won't work, we need access to the WebKitTab object...

File: qutebrowser/browser/webkit/network/networkmanager.py [hunk 2]
```diff
@@ -269,8 +269,8 @@ def clear_rejected_ssl_errors(self, url):
     def on_authentication_required(self, reply, authenticator):
         """Called when a website needs authentication."""
         netrc = False
-        if not hasattr(reply, "netrc_used"):
-            setattr(reply, "netrc_used", True)
+        if not self.data.netrc_used:
```
Comment (mjauvin) [inline qutebrowser/browser/webkit/network/networkmanager.py]: ok, this won't work, we need access to the WebKitTab object...

File: qutebrowser/browser/browsertab.py [hunk 3]
```diff
@@ -726,7 +724,12 @@ def _on_load_started(self):
         self._progress = 0
         self._has_ssl_errors = False
         self.data.viewing_source = False
-        self.data.netrc_used = False
+        if self.backend == usertypes.Backend.QtWebKit:
```
Comment (mjauvin) [inline qutebrowser/browser/browsertab.py]: the temporary attribute is cleared here for both backends.

File: qutebrowser/browser/browsertab.py [hunk 4]
```diff
@@ -764,7 +764,6 @@ def _on_load_finished(self, ok):
         self.load_finished.emit(ok)
         if not self.title():
             self.title_changed.emit(self.url().toDisplayString())
-
```
Comment (The-Compiler) [inline qutebrowser/browser/browsertab.py]: Unrelated change?

File: qutebrowser/browser/shared.py [hunk 5]
```diff
@@ -260,3 +262,29 @@ def get_user_stylesheet():
         css += '\nhtml > ::-webkit-scrollbar { width: 0px; height: 0px; }'
 
     return css
+
+
+def netrc_authentication(url, authenticator):
+    """performs authorization using netrc."""
```
Comment (The-Compiler) [inline qutebrowser/browser/shared.py]: This should be capitalized and in imperative mood ("Perform ..."), see [PEP 257](https://www.python.org/dev/peps/pep-0257/).

Also, please document the `url` and `authenticator` arguments as well as what the return value means.

File: qutebrowser/browser/shared.py [hunk 6]
```diff
@@ -260,3 +262,29 @@ def get_user_stylesheet():
...
+    if 'HOME' in os.environ:
+        # We'll get an OSError by netrc if 'HOME' isn't available in
+        # os.environ. We don't want to log that, so we prevent it
+        # altogether.
```
Comment (The-Compiler) [inline qutebrowser/browser/shared.py]: Now that this is a separate function, you can make this an early exit:

```python
if 'HOME' not in os.environ:
    # ...
    return False
```

File: qutebrowser/browser/webengine/webenginetab.py [hunk 7]
```diff
@@ -739,11 +739,15 @@ def _on_proxy_authentication_required(self, url, authenticator,
 
     @pyqtSlot(QUrl, 'QAuthenticator*')
     def _on_authentication_required(self, url, authenticator):
-        # FIXME:qtwebengine support .netrc
-        answer = shared.authentication_required(
-            url, authenticator, abort_on=[self.shutting_down,
-                                          self.load_started])
-        if answer is None:
+        netrc = None
+        if not hasattr(self, 'netrc_used'):
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webenginetab.py]: The reason `hasattr` was used in the old code is because the reply objects were created by Qt, so we couldn't control what attributes is has. Now that we can, there's no reason to use `hasattr`/`setattr`/`delattr` anymore - just set the attribute to `False` initially and test for truthyness.

Also, the attribute should be moved to `TabData` instead. It's a bit unfortunate that it's used only for QtWebEngine, but it's still the best solution, as there can be `QNetworkAccessManager` objects without a tab associated.

File: qutebrowser/browser/webengine/webenginetab.py [hunk 8]
```diff
@@ -739,11 +739,15 @@ def _on_proxy_authentication_required(self, url, authenticator,
...
+            setattr(self, 'netrc_used', True)
+            netrc = shared.netrc_authentication(url, authenticator)
```
Comment (The-Compiler) [inline qutebrowser/browser/webengine/webenginetab.py]: Maybe call this `netrc_success` or so?

File: qutebrowser/browser/webkit/network/networkmanager.py [hunk 9]
```diff
@@ -270,28 +268,11 @@ def clear_rejected_ssl_errors(self, url):
...
-                log.misc.exception("Error when parsing the netrc file")
-
-        if user is not None:
-            authenticator.setUser(user)
-            authenticator.setPassword(password)
-        else:
+        netrc = False
+        if not hasattr(self, 'netrc_used'):
+            setattr(self, 'netrc_used', True)
+            netrc = shared.netrc_authentication(reply.url(), authenticator)
```
Comment (The-Compiler) [inline qutebrowser/browser/webkit/network/networkmanager.py]: The comments from above about `netrc` and `settatr`/`hasattr` apply here as well. Also, it should be a private attribute here (`self._netrc_used`) as it's not used from the outside.

File: qutebrowser/browser/webkit/webkittab.py [hunk 10]
```diff
@@ -704,6 +704,13 @@ def user_agent(self):
         page = self._widget.page()
         return page.userAgentForUrl(self.url())
 
+    @pyqtSlot()
+    def _on_load_started(self):
+        super()._on_load_started()
+        obj = self.networkaccessmanager()
```
Comment (The-Compiler) [inline qutebrowser/browser/webkit/webkittab.py]: I'm sure you can think of a better name than `obj` here. `nam` for example :wink:

File: qutebrowser/browser/webkit/network/networkmanager.py (hunk 9 above)
Comment (mjauvin) [reply qutebrowser/browser/webkit/network/networkmanager.py]: It's actually accessed from the webkittab class in its _on_load_started() method, so I cannot make this private.

**Comment by The-Compiler** [comment]
Thanks! Merged with a few cleanups in 772f0025f18f46eddaa0649ce826d168f563a9a7 to rename `netrc` to `netrc_success` everywhere and not just at one place, and to set `netrc_used` unconditionally (no point in checking whether it's true, if it's false and we set it to false there's no harm done).
